pkg/debuginfo/store.go

```diff
@@ -214,72 +253,89 @@ func (s *Store) Upload(stream debuginfopb.DebugInfoService_UploadServer) error {
 	}
 
 	if found {
+		if hash != "" && metadataFile != nil {
+			if metadataFile.Hash == hash {
+				level.Debug(s.logger).Log("msg", "debug info already exists", "buildid", buildID)
+				return status.Error(codes.AlreadyExists, "debuginfo already exists")
+			}
+		}
+
 		objFile, err := s.fetchObjectFile(ctx, buildID)
 		if err != nil {
 			return status.Error(codes.Internal, err.Error())
 		}
 
-		if req.GetInfo().Hash != "" {
-			h, err := hash.File(objFile)
-			if err != nil {
-				return status.Error(codes.Internal, err.Error())
-			}
-
-			if h == req.GetInfo().Hash {
-				level.Debug(s.logger).Log("msg", "debug info already exists", "buildid", buildID)
-				return status.Error(codes.AlreadyExists, "debuginfo already exists")
+		if err := elfutils.ValidateFile(objFile); err != nil {
+			// Failed to validate. Mark the file as corrupted, and let the client try to upload it again.
+			if err := s.metadataManager.markAsCorrupted(ctx, buildID); err != nil {
+				level.Warn(s.logger).Log("msg", "failed to update metadata as corrupted", "err", err)
 			}
+			level.Error(s.logger).Log("msg", "failed to validate object file", "buildid", buildID)
+			// Client will retry.
+			return status.Error(codes.Internal, err.Error())
 		}
 
+		// Valid.
 		hasDWARF, err := elfutils.HasDWARF(objFile)
-		if err != nil && !errors.Is(err, io.EOF) {
-			var fe *elf.FormatError
-			if errors.As(err, &fe) {
-				// Ignore bad magic number if all zero
-				if !strings.Contains(fe.Error(), "bad magic number '[0 0 0 0]'") {
-					return status.Error(codes.Internal, err.Error())
-				}
-			} else {
-				return status.Error(codes.Internal, err.Error())
-			}
+		if err != nil {
+			level.Debug(s.logger).Log("msg", "failed to check for DWARF", "err", err)
 		}
 		if hasDWARF {
-			// We probably have the best version.
-			level.Debug(s.logger).Log("msg", "debug info with DWARF already exists", "buildid", buildID)
 			return status.Error(codes.AlreadyExists, "debuginfo already exists")
 		}
 	}
 
-	if err := s.metadataManager.update(ctx, buildID, metadataStateUploading); err != nil {
-		level.Error(s.logger).Log("msg", "failed to update metadata", "err", err)
+	if err := s.metadataManager.markAsUploading(ctx, buildID); err != nil {
+		err = fmt.Errorf("failed to update metadata before uploading: %w", err)
+		return status.Error(codes.Internal, err.Error())
 	}
 
-	// At this point we know that we still have a better version of the debug information file,
+	//nolint:forcetypeassert
+	b := s.pool.Get().(*bytes.Buffer)
+	// NOTICE: The ELF header is 52 or 64 bytes long for 32-bit and 64-bit binaries respectively.
+	buf := limitio.NewWriter(b, 64, true)
```

Comment:
This took me a couple of reads to understand, and I have a few questions about this.

1) Is the pooling really necessary? It seems that we'd only ever allocate max 64 bytes per upload.
2) I think for future me I would need a comment that this limitio Writer does not error when you attempt to write more than 64 bytes.
3) If I'm reading the logic right, then we're not actually validating the elf header until after we uploaded to the bucket. Isn't the whole point to prevent that?

Reply:
1. Yes, pooling is not strictly necessary here. Since this is a frequent recurring task, I've put the pool there. If you think it's overkill, I'm happy t remove it.
2. I should have commented and added an explanation. I will do that.
3. Yes, this happens after the fact. We're optimistically uploading directly to the bucket and if something goes wrong we mark it as corrupted, so it could be overwritten, in subsequent calls. We only want to make sure we don't read a corrupted file while symbolizing. I wanted to prevent any form of buffering this data on the server-side thus the optimistic writes directly to the object-store. I'm open to other suggestions.

Reply:
Oooooooo I see! Could you add what you wrote in 3 as a comment as well? :D 

Now it's all clear 💯 

(and yeah I think the pool is unnecessary)